endpoints/auth.go

```diff
@@ -55,56 +54,56 @@ var (
 		return time.Now().UTC()
 	}
 
-	// ContextFactory takes an in-flight HTTP request and creates a new
-	// context.
+	// AuthenticatorFactory creates a new authenticator.
 	//
 	// It is a variable on purpose. You can set it to a stub implementation
 	// in tests.
-	ContextFactory func(*http.Request) Context
+	AuthenticatorFactory func() Authenticator
 )
 
-// Context represents the context of an in-flight API request.
-// It embeds appengine.Context so you can use it with any other appengine/*
-// package methods.
-type Context interface {
-	appengine.Context
-
-	// HTTPRequest returns the request associated with this context.
-	HTTPRequest() *http.Request
-
-	// Namespace returns a replacement context that operates within the given namespace.
-	Namespace(name string) (Context, error)
-
+// An Authenticator can identify the current user.
+type Authenticator interface {
 	// CurrentOAuthClientID returns a clientID associated with the scope.
-	CurrentOAuthClientID(scope string) (string, error)
+	CurrentOAuthClientID(ctx context.Context, scope string) (string, error)
 
 	// CurrentOAuthUser returns a user of this request for the given scope.
 	// It caches OAuth info at the first call for future invocations.
 	//
 	// Returns an error if data for this scope is not available.
-	CurrentOAuthUser(scope string) (*user.User, error)
+	CurrentOAuthUser(ctx context.Context, scope string) (*user.User, error)
 }
 
-// NewContext returns a new context for an in-flight API (HTTP) request.
-func NewContext(req *http.Request) Context {
-	ctxsMu.Lock()
-	defer ctxsMu.Unlock()
-	c, ok := ctxs[req]
+// contextKey is used to store values on a context.
+type contextKey int
+
+// Context value keys.
+const (
+	requestKey contextKey = iota
```

Comment:
I'd feel more comfortable if the zero value for contextKey was not a valid key.

```
const (
    invalidKey contextKey = iota
    requestKey
    authenticatorKey
)
```

Or maybe use strings instead of ints?